File: life-journey-frontend/src/lib/indexing.ts
/**
 * Client-side SEO indexing utilities.
 * Gebruikt door de /api/test-indexing route en eventuele directe pings vanuit Next.js.
 *
 * Voor productie-publicaties doet de FastAPI backend de indexering server-to-server.
 */

export async function pingIndexNow(urls: string[]): Promise<void> {
  const key = process.env.INDEXNOW_KEY;
  if (!key) {
    console.debug("[indexing] INDEXNOW_KEY not set, skipping IndexNow");
    return;
  }

  if (urls.length === 0) return;

  const host = new URL(urls[0]).hostname;

  try {
    const res = await fetch("https://api.indexnow.org/indexnow", {
      method: "POST",
      headers: { "Content-Type": "application/json; charset=utf-8" },
      body: JSON.stringify({
        host,
        key,
        keyLocation: `${process.env.NEXT_PUBLIC_SITE_URL}/${key}.txt`,
        urlList: urls,
      }),
    });

    if (res.ok || res.status === 202) {
      console.log(`[indexing] IndexNow OK — ${urls.length} URL(s) submitted`);
    } else {
      console.warn(`[indexing] IndexNow HTTP ${res.status}`);
    }
  } catch (err) {
    console.warn("[indexing] IndexNow error (non-critical):", err);
  }
}

export async function pingGoogleIndexingAPI(url: string): Promise<void> {
  const saJson = process.env.GOOGLE_SERVICE_ACCOUNT_JSON;
  if (!saJson) {
    console.debug("[indexing] GOOGLE_SERVICE_ACCOUNT_JSON not set, skipping Google Indexing");
    return;
  }

  try {
    const token = await getGoogleAccessToken(saJson);
    if (!token) return;

    const res = await fetch(
      "https://indexing.googleapis.com/v3/urlNotifications:publish",
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${token}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ url, type: "URL_UPDATED" }),
      }
    );

    if (res.ok) {
      console.log(`[indexing] Google Indexing OK — ${url}`);
    } else {
      console.warn(`[indexing] Google Indexing HTTP ${res.status}`);
    }
  } catch (err) {
    console.warn("[indexing] Google Indexing error (non-critical):", err);
  }
}

async function getGoogleAccessToken(saJson: string): Promise<string | null> {
  try {
    const sa = JSON.parse(saJson);
    const now = Math.floor(Date.now() / 1000);

    const header = { alg: "RS256", typ: "JWT" };
    const claim = {
      iss: sa.client_email,
      scope: "https://www.googleapis.com/auth/indexing",
      aud: "https://oauth2.googleapis.com/token",
      exp: now + 3600,
      iat: now,
    };

    const encode = (obj: object) =>
      Buffer.from(JSON.stringify(obj))
        .toString("base64")
        .replace(/=/g, "")
        .replace(/\+/g, "-")
        .replace(/\//g, "_");

    const headerB64 = encode(header);
    const claimB64 = encode(claim);
    const signingInput = `${headerB64}.${claimB64}`;

    // Node.js crypto for RSA signing (server-side only)
    const { createSign } = await import("crypto");
    const sign = createSign("RSA-SHA256");
    sign.update(signingInput);
    const signature = sign
      .sign(sa.private_key)
      .toString("base64")
      .replace(/=/g, "")
      .replace(/\+/g, "-")
      .replace(/\//g, "_");

    const jwt = `${signingInput}.${signature}`;

    const tokenRes = await fetch("https://oauth2.googleapis.com/token", {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({
        grant_type: "urn:ietf:params:oauth:grant-type:jwt-bearer",
        assertion: jwt,
      }),
    });

    const data = await tokenRes.json();
    return data.access_token ?? null;
  } catch (err) {
    console.warn("[indexing] Google token error:", err);
    return null;
  }
}

File: life-journey-backend/app/services/indexing.py
"""
SEO indexing service — pings IndexNow en Google Indexing API na publicatie.

Beide calls zijn non-blocking: een fout breekt de publish-flow NOOIT.
"""
import json
import time
from typing import Optional

import httpx
from loguru import logger

from app.core.config import settings


async def ping_index_now(urls: list[str]) -> None:
    """Stuur URLs naar IndexNow (Bing/Yandex/Google-compatible)."""
    if not settings.indexnow_key:
        logger.debug("INDEXNOW_KEY niet ingesteld, IndexNow overgeslagen")
        return

    if not urls:
        return

    payload = {
        "host": _extract_host(urls[0]),
        "key": settings.indexnow_key,
        "keyLocation": f"{settings.site_url}/{settings.indexnow_key}.txt",
        "urlList": urls,
    }

    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            resp = await client.post(
                "https://api.indexnow.org/indexnow",
                json=payload,
                headers={"Content-Type": "application/json; charset=utf-8"},
            )
            if resp.status_code in (200, 202):
                logger.info(f"IndexNow OK — {len(urls)} URL(s) ingediend")
            else:
                logger.warning(f"IndexNow HTTP {resp.status_code}: {resp.text[:200]}")
    except Exception as exc:
        logger.warning(f"IndexNow fout (niet kritiek): {exc}")


async def ping_google_indexing_api(url: str) -> None:
    """Meld een URL aan bij de Google Indexing API via Service Account."""
    if not settings.google_service_account_json:
        logger.debug("GOOGLE_SERVICE_ACCOUNT_JSON niet ingesteld, Google Indexing overgeslagen")
        return

    try:
        token = await _get_google_access_token()
        if not token:
            return

        async with httpx.AsyncClient(timeout=10.0) as client:
            resp = await client.post(
                "https://indexing.googleapis.com/v3/urlNotifications:publish",
                json={"url": url, "type": "URL_UPDATED"},
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json",
                },
            )
            if resp.status_code == 200:
                logger.info(f"Google Indexing OK — {url}")
            else:
                logger.warning(f"Google Indexing HTTP {resp.status_code}: {resp.text[:200]}")
    except Exception as exc:
        logger.warning(f"Google Indexing fout (niet kritiek): {exc}")


async def _get_google_access_token() -> Optional[str]:
    """Genereer een OAuth2 access token via Service Account JWT."""
    try:
        sa = json.loads(settings.google_service_account_json)
    except Exception:
        logger.warning("GOOGLE_SERVICE_ACCOUNT_JSON is geen geldige JSON")
        return None

    try:
        import importlib
        jwt_lib = importlib.import_module("jwt")
    except ImportError:
        logger.warning("PyJWT niet beschikbaar voor Google service account auth")
        return None

    now = int(time.time())
    claim = {
        "iss": sa["client_email"],
        "scope": "https://www.googleapis.com/auth/indexing",
        "aud": "https://oauth2.googleapis.com/token",
        "exp": now + 3600,
        "iat": now,
    }

    try:
        signed_jwt = jwt_lib.encode(claim, sa["private_key"], algorithm="RS256")
        async with httpx.AsyncClient(timeout=10.0) as client:
            resp = await client.post(
                "https://oauth2.googleapis.com/token",
                data={
                    "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
                    "assertion": signed_jwt,
                },
            )
            data = resp.json()
            return data.get("access_token")
    except Exception as exc:
        logger.warning(f"Google OAuth token fout: {exc}")
        return None


def _extract_host(url: str) -> str:
    try:
        from urllib.parse import urlparse
        return urlparse(url).netloc
    except Exception:
        return ""
